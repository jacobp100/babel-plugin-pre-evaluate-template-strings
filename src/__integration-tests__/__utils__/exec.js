/* @flow */

import childProcess from 'child_process';
import path from 'path';

export function transpile(
  source: string,
  pluginOptions: any = { cache: false, extract: false },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_TRANSPILE_BIN = path.join(__dirname, './transpile.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_TRANSPILE_BIN} '${JSON.stringify(
      source
    )}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [transpiled, rawStyles] = stdout.toString().split('\n');
  const styles = JSON.parse(rawStyles || '{}');
  return {
    code: JSON.parse(transpiled || '""'),
    styles,
    getCSSForClassName(className: string) {
      return styles[`.${className}`];
    },
  };
}

export function extract(
  source: string,
  pluginOptions: any = { cache: false, extract: true },
  babelOptions: any = { filename: 'test.js' }
) {
  const PATH_TO_EXTRACT_BIN = path.join(__dirname, './extract.js');

  const stdout = childProcess.execSync(
    `node ${PATH_TO_EXTRACT_BIN} '${JSON.stringify(source)}' '${JSON.stringify({
      pluginOptions,
      babelOptions,
    })}'`,
    { stdio: 'pipe' }
  );

  const [data, filenames, transpiled] = stdout.toString().split('\n');
  return {
    data: JSON.parse(data || '""'),
    filenames: JSON.parse(filenames || '[]'),
    transpiled: JSON.parse(transpiled || '""'),
  };
}
