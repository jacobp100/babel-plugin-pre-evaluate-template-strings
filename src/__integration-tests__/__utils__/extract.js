const fs = require('fs');

const writeFileSync = fs.writeFileSync;

let buffer = '';
const filenames = [];

fs.writeFileSync = (filename, data, opts) => {
  if (/\.css$/.test(filename)) {
    filenames.push(filename);
    buffer += data;
    return;
  }
  writeFileSync(filename, data, opts);
};

const transpile = require('./transpile');

const code = transpile(
  JSON.parse(process.argv[2]),
  JSON.parse(process.argv[3])
);

console.log(JSON.stringify(buffer));
console.log(JSON.stringify(filenames));
console.log(JSON.stringify(code || ''));
