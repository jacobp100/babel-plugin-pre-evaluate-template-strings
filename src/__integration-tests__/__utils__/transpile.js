const babel = require('babel-core');
const path = require('path');
const dedent = require('dedent');

const CSS_PATH = require.resolve(path.join(process.cwd(), 'build/css'));
const SHEET_PATH = require.resolve(path.join(process.cwd(), 'build/sheet'));
const PREVAL_PLUGIN_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/preval-extract')
);
const MODULE_SYSTEM_PATH = require.resolve(
  path.join(process.cwd(), 'build/babel/lib/moduleSystem')
);

function transpile(source, options = { pluginOptions: {}, babelOptions: {} }) {
  return babel.transform(
    dedent`
      import css from '${CSS_PATH}';

      ${source}
    `,
    Object.assign(
      {
        presets: ['env', 'stage-2', 'react'],
        plugins: [[PREVAL_PLUGIN_PATH, options.pluginOptions]],
        sourceMaps: false,
      },
      options.babelOptions
    )
  ).code;
}

module.exports = transpile;

if (require.main && module.filename === require.main.filename) {
  const code = transpile(
    JSON.parse(process.argv[2]),
    JSON.parse(process.argv[3])
  );

  if (code) {
    console.log(JSON.stringify(code));

    // eslint-disable-next-line global-require
    const getCachedModule = require(MODULE_SYSTEM_PATH).getCachedModule;
    const sheet = getCachedModule(SHEET_PATH);
    const styles = sheet
      ? getCachedModule(SHEET_PATH).exports.default.styles()
      : {};
    console.log(JSON.stringify(styles || ''));
  }
}
