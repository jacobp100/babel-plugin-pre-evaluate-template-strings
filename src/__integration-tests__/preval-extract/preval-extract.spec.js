/* eslint-disable no-template-curly-in-string */
/* @flow */

import dedent from 'dedent';
import { transpile } from '../__utils__/exec';

describe('preval-extract babel plugin', () => {
  it('should not process tagged template if tag is not "css"', () => {
    const { code } = transpile(dedent`
    const header = \`
      font-size: 3em;
    \`;
    `);
    expect(code).toMatch('font-size: 3em;');
    expect(code).toMatchSnapshot();
  });

  it('should not process file if it has "linaria-preval" comment', () => {
    const { code } = transpile(dedent`
    /* linaria-preval */
    const header = css\`
      font-size: ${'${2 + 1}'}em;
    \`;
    `);
    expect(code).not.toMatch('font-size: 3em;');

    const { code: codeAlt } = transpile(dedent`
    const header = css\`
      font-size: ${'${2 + 1}'}em;
    \`;
    /* linaria-preval */
    `);
    expect(codeAlt).not.toMatch('font-size: 3em;');
  });

  it('should create classname for "css" tagged template literal', () => {
    const { code, getCSSForClassName } = transpile(dedent`
    const header = css\`
      font-size: 3em;
    \`;
    `);

    const match = /header = \/\*.+\*\/'(_header__[a-z0-9]+)'/g.exec(code);
    expect(match).not.toBeNull();
    const css = getCSSForClassName(match[1]);
    expect(css).toMatch('font-size: 3em');
    expect(css).toMatchSnapshot();
  });

  it('should create classname for "css.named()" tagged template literal', () => {
    const { code, getCSSForClassName } = transpile(dedent`
    const header = css.named("my-header")\`
      font-size: 3em;
    \`;
    `);

    const match = /header = \/\*.+\*\/"(my-header__[a-z0-9]+)"/g.exec(code);
    expect(match).not.toBeNull();
    const css = getCSSForClassName(match[1]);
    expect(css).toMatch('font-size: 3em');
    expect(css).toMatchSnapshot();
  });

  it('should create classnames for multiple "css" tagged template literal', () => {
    const { code, getCSSForClassName } = transpile(dedent`
    const header = css\`
      font-size: ${'${2 + 1}'}em;
    \`;

    const body = css\`
      border-radius: ${'${2 + 2}'}px;
    \`;
    `);

    const headerMatch = /header = \/\*.+\*\/'(_header__[a-z0-9]+)'/g.exec(code);
    const bodyMatch = /body = \/\*.+\*\/'(_body__[a-z0-9]+)'/g.exec(code);
    expect(headerMatch).not.toBeNull();
    expect(bodyMatch).not.toBeNull();
    const headerStyles = getCSSForClassName(headerMatch[1]);
    const bodyStyles = getCSSForClassName(bodyMatch[1]);
    expect(headerStyles).toMatch('font-size: 3em');
    expect(headerStyles).toMatchSnapshot();
    expect(bodyStyles).toMatch('border-radius: 4px');
    expect(bodyStyles).toMatchSnapshot();
  });

  it('should create classname for non-top-level "css" tagged template literal', () => {
    const { code, getCSSForClassName } = transpile(dedent`
    const defaults = {
      fontSize: "3em",
    };
    function render() {
      const header = css\`
        font-size: ${'${defaults.fontSize}'};
      \`;
    }
    `);

    const match = /header = \/\*.+\*\/"(_header__[a-z0-9]+)"/g.exec(code);
    expect(match).not.toBeNull();
    const css = getCSSForClassName(match[1]);
    expect(css).toMatch('font-size: 3em');
    expect(css).toMatchSnapshot();
  });

  it('should preval const and let without transpilation to var', () => {
    const { code, getCSSForClassName } = transpile(
      dedent`
      const size = 3;
      let color = "#ffffff";

      const header = css\`
        font-size: ${'${size}'}em;
        color: ${'${color}'};
      \`;
      `,
      undefined,
      {
        presets: ['stage-2'],
        filename: 'test.js',
        babelrc: false,
      }
    );

    const match = /header = \/\*.+\*\/"(_header__[a-z0-9]+)"/g.exec(code);
    expect(match).not.toBeNull();
    const css = getCSSForClassName(match[1]);
    expect(css).toMatch('font-size: 3em');
    expect(css).toMatch('color: #ffffff');
    expect(code).toMatchSnapshot();
  });

  it('should preval css with classname from another prevaled css', () => {
    const { code, getCSSForClassName } = transpile(dedent`
    const title = css\`
    color: blue;
    \`;

    const container = css\`
    padding: 2rem;

    .${'${title}'} {
      margin-bottom: 1rem;
    }
    \`;
    `);

    const titleMatch = /title = \/\*.+\*\/'(_title__[a-z0-9]+)'/g.exec(code);
    expect(titleMatch).not.toBeNull();

    const containerMatch = /container = \/\*.+\*\/'(_container__[a-z0-9]+)'/g.exec(
      code
    );
    expect(containerMatch).not.toBeNull();

    const css = getCSSForClassName(containerMatch[1]);
    expect(css).toMatch(titleMatch[1]);
    expect(css).toMatchSnapshot();
  });
});
