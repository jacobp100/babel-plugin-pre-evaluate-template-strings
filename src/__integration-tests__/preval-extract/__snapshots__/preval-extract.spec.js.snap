// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval-extract babel plugin should create classname for "css" tagged template literal 1`] = `"font-size: 3em;"`;

exports[`preval-extract babel plugin should create classname for "css.named()" tagged template literal 1`] = `"font-size: 3em;"`;

exports[`preval-extract babel plugin should create classname for non-top-level "css" tagged template literal 1`] = `"font-size: 3em;"`;

exports[`preval-extract babel plugin should create classnames for multiple "css" tagged template literal 1`] = `"font-size: 3em;"`;

exports[`preval-extract babel plugin should create classnames for multiple "css" tagged template literal 2`] = `"border-radius: 4px;"`;

exports[`preval-extract babel plugin should not process tagged template if tag is not "css" 1`] = `
'use strict';

var _css = require('<<CWD>>/build/css.js');

var _css2 = _interopRequireDefault(_css);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var header = '\\nfont-size: 3em;\\n';
`;

exports[`preval-extract babel plugin should preval const and let without transpilation to var 1`] = `
import css from '<<CWD>>/build/css.js';

const size = 3;
let color = "#ffffff";

const header = /*linaria-output*/"_header__zkmz8m";
`;

exports[`preval-extract babel plugin should preval css with classname from another prevaled css 1`] = `
"padding: 2rem;

._title__zkmz8m {
margin-bottom: 1rem;
}"
`;
